feat: Support configuring rule.Reason as a Sprintf format string

jessehu opened the pull request:
So we can define one templated Reason to generate many Reasons, e.g. generating Reasons for Nvidia GPU Xid errors in dmesg log.

Close #1067

jessehu committed:
feat: Support configuring rule.Reason as a Sprintf format string

So we can define one templated Reason to generate many Reasons, e.g.
generating Reasons for Nvidia GPU Xid errors in dmesg log.

jessehu commented:
/retest

jessehu commented:
Thanks @hakman for review! Looks like [pull-npd-e2e-test](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/node-problem-detector/1068/pull-npd-e2e-test/1935506399140253696) failed twice. Not sure if it's related to this PR ?
![图片](https://github.com/user-attachments/assets/91e27324-a0f9-489b-97c8-8bb4d3abea9b)

hakman commented:
> Thanks @hakman for review! Looks like [pull-npd-e2e-test](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/node-problem-detector/1068/pull-npd-e2e-test/1935506399140253696) failed twice. Not sure if it's related to this PR ? 

Unrelated to the PR, please ignore.

jessehu commented:
/retest

jessehu commented:
Thanks @hakman. The test passed now. Please take a look again.

jessehu commented:
/retest

googs1025 commented on pkg/systemlogmonitor/log_monitor.go:
```diff
@@ -169,12 +174,31 @@ func (l *logMonitor) generateStatus(logs []*systemlogtypes.Log, rule systemlogty
 	var changedConditions []*types.Condition
+
+	// Support configuring rule.Reason as a Sprintf format string and formatting it with the matched capturing groups in rule.Pattern.
+	re := regexp.MustCompile(rule.Pattern)
+	matches := re.FindStringSubmatch(message)
```
Also, is it possible that re.FindStringSubmatch(message) returns nil?

jessehu replied:
@googs1025 NPD validates all rule.Pattern at the beginning, so using MustCompile() here is OK.
https://github.com/jessehu/node-problem-detector/blob/411a1baa0b52b1bd57aa17a414efbc56bb5c3d80/pkg/systemlogmonitor/log_monitor.go#L76-L80
https://github.com/jessehu/node-problem-detector/blob/411a1baa0b52b1bd57aa17a414efbc56bb5c3d80/pkg/systemlogmonitor/config.go#L62-L71

googs1025 replied:
got this 😄

jessehu replied:
> Also, is it possible that re.FindStringSubmatch(message) returns nil?

@googs1025 the caller of generateStatus() already ensures it will match: https://github.com/jessehu/node-problem-detector/blob/411a1baa0b52b1bd57aa17a414efbc56bb5c3d80/pkg/systemlogmonitor/log_monitor.go#L152-L183.  Also `if len(matches) > 1` will be false even if `matches` is a nil []string.

jessehu commented:
/retest

jessehu committed:
fix: Allow capturing groups in rule.Pattern with non-templated rule.Reason

hakman commented:
/remove-lifecycle rotten

DigitalVeer commented on pkg/systemlogmonitor/log_monitor.go:
```diff
@@ -170,12 +175,33 @@ func (l *logMonitor) generateStatus(logs []*systemlogtypes.Log, rule systemlogty
+
+	reason := rule.Reason
+	// Support configuring rule.Reason as a Sprintf format string and formatting it with the matched capturing groups in rule.Pattern.
+	if strings.Contains(reason, "%") {
```
This turns every `%` into a template. 

`"Disk90%"` becomes --> `"Disk90%!(NOVERB)"` and can cause us to fail to match.

jessehu replied:
@DigitalVeer An invalid template string like "Disk90%" will cause NPD process to panic when launching, so the user can find it out easily rather than not matching silently.

DigitalVeer replied:
Thanks, Jesse. I tried to reproduce the panic with a new filelog monitor:
>
> ```json
> {
>   "plugin": "filelog",
>   "pluginConfig": {
>     "timestamp": "^(\\S+)",
>     "message": "^\\S+ (.*)$",
>     "timestampFormat": "2006-01-02T15:04:05Z07:00"
>   },
>   "logPath": "/repro/pr_1068_literal_percent.log",
>   "lookback": "24h",
>   "bufferSize": 1,
>   "source": "literal-percent-e2e",
>   "metricsReporting": true,
>   "conditions": [],
>   "rules": [
>     {
>       "type": "temporary",
>       "reason": "Disk90%",
>       "pattern": "disk usage reached 90%"
>     }
>   ]
> }
> ```


I wrote this line to the log file:

> ```text
> 2026-07-15T09:58:59Z disk usage reached 90%
> ```

NPD started normally. After filelog read and matched the line, NPD logged:

> ```text
> Got wrong string "Disk90%!(NOVERB)" for reason "Disk90%" with pattern "disk usage reached 90%"
> ```

The process remained alive, but NPD generated no status or event. I could not reproduce either a launch or runtime panic. The match is dropped at runtime instead.

If the goal is to surface bad reason strings easily one option is validating them in [ValidateRules](https://github.com/kubernetes/node-problem-detector/blob/6ccf48be371611f1be4746444414da4eb1d7f48a/pkg/systemlogmonitor/config.go#L63) which would let us catch them at startup.

jessehu replied:
@DigitalVeer I see. `"pattern": "disk usage reached 90%"` is a valid regex pattern，so NPD doesn't panic in [ValidateRules](https://github.com/kubernetes/node-problem-detector/blob/6ccf48be371611f1be4746444414da4eb1d7f48a/pkg/systemlogmonitor/config.go#L63), but `"reason": "Disk90%"` is not a valid Sprintf templated string. I have added validation for reason in ValidateRules().

jessehu committed:
feat: Validate rule.Reason as a Sprintf format string in ValidateRules

Dry-run fmt.Sprintf against the pattern's capturing group count so
misconfigured Reason templates (wrong verb count or unsupported verbs
for string args) fail at config load time instead of being silently
dropped at runtime.

DigitalVeer commented:
/retest
